src/test_fixture/world.rs

```diff
@@ -108,15 +136,54 @@ impl TestWorld {
     #[must_use]
     pub fn contexts<F: Field>(&self) -> [SemiHonestContext<'_, F>; 3] {
         let execution = self.executions.fetch_add(1, Ordering::Release);
-        let run = format!("run-{execution}");
-        zip(Role::all(), zip(&self.participants, &self.gateways))
+        zip(Role::all(), zip(&self.participants, &*self.gateways))
             .map(|(role, (participant, gateway))| {
-                SemiHonestContext::new(*role, participant, gateway).narrow(&run)
+                SemiHonestContext::new(*role, participant, gateway)
+                    .narrow(&Self::execution_step(execution))
             })
             .collect::<Vec<_>>()
             .try_into()
             .unwrap()
     }
+
+    #[must_use]
+    pub fn metrics_snapshot(&self) -> Metrics {
+        self.metrics_handle.snapshot()
+    }
+
+    #[must_use]
+    pub fn execution_step(execution: usize) -> impl Substep {
+        format!("run-{execution}")
+    }
+
+    pub fn gateway(&self, role: Role) -> &Gateway {
+        &self.gateways[role]
+    }
+
+    #[cfg(not(feature = "shuttle"))]
+    pub async fn join(mut self) {
```

Comment:
Do we need to add `world.join().await` to a bunch of tests?  Or is there something we can do with tokio that we can use.  I couldn't see anything obvious.